hazelcast-jet-core/src/main/java/com/hazelcast/jet/Observable.java

```diff
@@ -0,0 +1,139 @@
+ * their {@link Observable#destroy()} method, whenever they are no longer
+ * needed. Automatic clean-up shouldn't be relied upon. Even if a client
+ * that has requested the {@link Observable} happens to crash before
+ * managing to do clean-up, it is still possible to manually destroy
+ * the backing {@link Ringbuffer}s by obtaining a new, identically named
+ * {@link Observable} and calling {@link Observable#destroy()} on that.
+ * <p>
+ * It is possible to use the same {@link Observable} for multiple jobs, or
+ * even use the same {@link Observable} multiple times in the same job (by
+ * defining {@link com.hazelcast.jet.pipeline.Sinks#observable(String)
+ * observable type sinks} with the same name). If done so one should be aware
+ * that there will be parallel streams of events which can be intermingled
+ * with eachother in all kinds of unexpected ways. It is not possible to
+ * tell which events originate from which sink. Newly registered
+ * {@link Observer}s seeing all events still in the {@link Ringbuffer}
+ * further exacerbates this problem so we recommend that an
+ * {@link Observable} with a certain name be used in a single sink of a
+ * single job.
+ *
+ * @param <T> type of the values in the sequence
+ */
+public interface Observable<T> extends Iterable<T> {
+
+    /**
+     * Returns the name identifying this particular observable.
+     * @return name of observable
+     */
+    String name();
+
+    /**
+     * Register an instance of {@link Observer} to be notified about any
+     * future and past events (to the extent that these haven't yet been
+     * overwritten in the backing {@link Ringbuffer}).
+     *
+     * @return registration ID associated with the added {@link Observer},
+     * can be used to remove the {@link Observer} later
+     */
+    UUID addObserver(@Nonnull Observer<T> observer);
+
+    /**
+     * Removes previously added {@link Observer}s, identified by their
+     * assigned registration IDs. Removed {@link Observer}s will
+     * not get notified about further events.
+     */
+    void removeObserver(UUID registrationId);
+
+    /**
+     * Non-thread safe iterator that can block while waiting for additional
+     * events.
+     * <p>
+     * Is backed by a blocking list which stores all event until they
+     * wait to be iterated over. Completion and error event also get put
+     * into the list, thus they are serialized with data events.
+     * Iteration will continue until either a completion or error event
+     * is encountered.
+     * <p>
+     * When no further event are available, but neither completion nor
+     * error has been encountered, the {@link Iterator#hasNext() hasNext()}
+     * method will block.
+     * <p>
+     * When a failure event is encountered the {@link Iterator#hasNext()
+     * hasNext()} call throws it as an Exception.
+     */
+    @Override
+    @Nonnull
+    default Iterator<T> iterator() {
```

Comment:
@cangencer had an idea to add `stream()` too. something along these lines should work:
```java
@Nonnull
default Stream<T> stream() {
  Spliterator<T> spliterator = Spliterators.spliteratorUnknownSize(iterator(), 0);
  return StreamSupport.stream(spliterator, false);
}
```

This allows you to write code like this:
```
Observable<MyResultObject> results = jet.getObservable("results");
List<MyResultObject> collected = results.stream().collect(toList());
System.out.println(collected);
```
The big question is whether it's a good idea to even attempt to have it all in the first release. Especially given this is useful for batch jobs only.

Reply:
So should I add it or not?

Reply:
I don't know. I am leaning towards not introducing it. Not in 4.0 release anyway. 
We can always add it later if needed/requested.

Reply:
I'm for introducing it, since it's just a minor wrapper over iterator and makes it convenient for example to collect the results into a map.

Reply:
My opinion is not really strong here, so I am good with including it too.